### DISCUSSION TITLE
tutorials/shaders/shader_reference/canvas_item_shader

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#174] [User-contributed notes]
# tutorials/shaders/shader_reference/canvas_item_shader

CanvasItem shaders are used to draw all 2D elements in Godot. These include all nodes that inherit from CanvasItems, and all GUI elements. CanvasItem shaders contain less built-in variables and fun...

https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/canvas_item_shader.html

**Comment by thygrrr** [comment]
Note! 

Unlike `shader_type spatial;` shaders, `shader_type canvas_item;` shaders have a different name for their fragment function - it's `fragment`, not `frag`.

```glsl
void fragment()
{
	COLOR = texture(TEXTURE, UV);
}
```

Sadly, this documentation page omits this. When creating a `void frag()`, it will not have the builtins you may expect.

**Reply by thygrrr** [reply]
A minimal canvas_item shader to start working with would thus be:
```glsl
shader_type canvas_item;

void fragment()
{
	COLOR = texture(TEXTURE, UV);
}
```

**Reply by Calinou** [reply]
It's `fragment()` in both `spatial` and `canvas_item` shaders (in both 3.x and 4.x). I've never seen `frag()` functions as built-in functions in Godot.

**Comment by vanBerloDevelopments** [comment]
Few useful tips I want to share about canvas_item shaders:

1. If you want to work with `blend_premul_alpha`, make sure to enable `premult_alpha` in the import settings before using the render mode `blend_premul_alpha`. The editor will do the work for you, so you don't have to import your images premultiplied!

2. To perform calculations with world space coordinates in screen space, or convert different coordinate spaces (e.g., world space to canvas space) and access those coordinates in the fragment shader, use varyings:

```glsl
shader_type canvas_item;

uniform vec2 world_space_coordinates;
varying vec2 transformed_world_to_canvas_coordinates;

void vertex() {
    transformed_world_to_canvas_coordinates = (CANVAS_MATRIX * vec4(world_space_coordinates, 0.0, 1.0)).xy;
}

void fragment() {
    // Access here
   transformed_world_to_canvas_coordinates
}

void light() {
    // Or here
   transformed_world_to_canvas_coordinates
}
```

3. For a blur effect, you can use mipmaps. Enable mipmaps in your import settings. Then sample a texture using `textureLod()` instead of `texture()`:

```glsl
vec4 blurred_texture = textureLod(TEXTURE, UV, 2.0);
```
The higher the last parameter, the blurrier the texture becomes.

4. Scaling a texture using UV mapping:
   * To scale 'up', multiply UV coordinates by a value lower than 1.
   * To scale 'down', multiply UV coordinates by a value higher than 1.
   
   For visual examples, see [this video tutorial](https://www.youtube.com/watch?v=tbcgEIvJlrI&t=109s).

5. Moving (translating) a texture using UV mapping:
   * To translate left: Add a positive value to UV.x
   * To translate right: Subtract a positive value from UV.x
   * To translate up: Add a positive value to UV.y
   * To translate down: Subtract a positive value from UV.y

   Example:
   ```glsl
   vec4 sampled_to_left = texture(TEXTURE, vec2(UV.x + 0.1, UV.y));
   ```

   For visual examples, see [this video tutorial](https://www.youtube.com/watch?v=tbcgEIvJlrI&t=109s).

Hope this helps! For those interested in mastering 2D shaders (no prior knowledge required), consider the course at [godot2dshaders.com](https://www.godot2dshaders.com) or visit [Godot Shader Boy on YouTube](https://www.youtube.com/@godotShaderBoy).

**Comment by abrusle** [comment]
Quick tip to make the canvas item's UVs relative to the region rect:
```glsl
void fragment() {
    vec2 region_uv = (UV - REGION_RECT.xy) * (1.0 / REGION_RECT.zw);
}
```

**Comment by Hydrcat** [comment]
If `SCREEN PIXEL_SIZE` is used in your shader , please note that it may result in different performance in the editor and running.You could find more in this->https://github.com/godotengine/godot/issues/92311

**Comment by snekky-studios** [comment]
For anyone trying to apply a fragment shader to a an atlastexture or region of a normal texture, here is a minimal example of how to use REGION_RECT:

void fragment() {
	vec2 uv_scaled = (UV - REGION_RECT.xy) / REGION_RECT.zw;
    // now use uv_scaled anywhere you would have used UV
}

I could not find an example anywhere else of how to actually use this.

**Reply by snekky-studios** [reply]
well that's embarrassing, a guy two comments above mine said the exact same thing.

**Reply by sprobus** [reply]
Your comment was actually very helpful to me; the other comment didn't mention anything about Atlas Textures, so I skipped by it without actually reading it.

For anyone else trying to get a gradient working on label text, the font is apparently stored as an Atlas Texture, so you can use the snippet above to get actually useful uv coordinates. It took me a long time to figure out why using the regular uv coordinates produced bizarre results where the shading of letters seemed to have no discernable pattern.
